flatten compressed cluster output

davidrohr opened the pull request:
@ktf @matthiasrichter @shahor02 : This is an attempt to flatten the compressed cluster output from the source. It is huge because it also includes #3499, will clean up once the other one is merged.
The flat structure itself and writing to it seems to be working, but I am having trouble sending it via DPL.
The output buffer is allocated as
```
size_t bufferSize = 2048ul * 1024 * 1024; // TODO: Just allocated some large buffer for now, should estimate this correctly;
auto bufferCompressedClusters = pc.outputs().make<std::vector<char>>(Output{gDataOriginTPC, "COMPCLUSTERS", 0}, bufferSize);
```
in line 550, CATrackerSpec.cxx, and retrieved as
```
auto tmp2 = pc.inputs().get<gsl::span<char>>("input");
```
in line 51, EntropyEncoderSpec.cxx.

I have 2 problems:
- The receiver receives always the buffer with the size as originally allocated. Using `bufferCompressedClusters.resize(...)` or `bufferCompressedClusters.push_back(...)` or `bufferCompressedClusters.emplace_back(...)` works and modifies the buffer on the sender side, but the receiver always receives a buffer of size `bufferSize`.
- The receiver receives only zeroes, irrespective of what I write into the buffer on the sender side.

I added some debug output:
On the sender:
```
CompressedClustersFlat* tmp = (CompressedClustersFlat*) bufferCompressedClusters.data();
printf("XXXXXXX Test %d %lld\n", (int) tmp->totalDataSize, (long long int) bufferCompressedClusters.size());
```
and on the receive:
```
CompressedClustersFlat* tmp = (CompressedClustersFlat*) tmp2.data();
printf("XXXXXXX Test %d %lld\n", (int) tmp->totalDataSize, (long long int) tmp2.size());
CompressedClusters compressed(*tmp);
LOG(INFO) << "CCCCC input data with " << compressed.nTracks << " track(s) and " << compressed.nAttachedClusters << " attached clusters";
```
But the printouts I see are:
```
[4686:tpc-tracker]: XXXXXXX Test 22128 22128
[4720:tpc-entropy-encoder]: XXXXXXX Test 0 2147483648
[4720:tpc-entropy-encoder]: [12:45:08][INFO] CCCCC input data with 0 track(s) and 0 attached clusters
```
So the XXXXXXX... outputs are not the same. I am sorry but I don't manage to understand what I am doing wrong. Could you have a look, or advise how I should allocate / send / receive the buffer.

shahor02 commented on Detectors/TPC/workflow/src/CATrackerSpec.cxx:
```diff
@@ -538,7 +539,12 @@ DataProcessorSpec getCATrackerSpec(ca::Config const& specconfig, std::vector<int
-      int retVal = tracker->runTracking(&ptrs);
+      GPUInterfaceOutputs outputRegions;
+      size_t bufferSize = 2048ul * 1024 * 1024; // TODO: Just allocated some large buffer for now, should estimate this correctly;
+      auto bufferCompressedClusters = pc.outputs().make<std::vector<char>>(Output{gDataOriginTPC, "COMPCLUSTERS", 0}, bufferSize);
```
Looks like you are creating a local copy here instead of taking the ref., should be:
``auto& bufferCompressedClusters = pc.outputs().make<std::vector<char>>...``

davidrohr replied:
That would explain all my problems, thanks, will check and report back.

davidrohr commented:
@shahor02 @matthiasrichter @MichaelLettrich 
The new version works for me (thx to @shahor02 ) and I rebased onto master.
I split the `CompressedClusters` struct in 3 variants:
- The former ROOT-serializable `CompressedClusters` has been renamed to `CompressedClustersROOT`.
- The new `CompressedClusters` is just for using it, but not for serialization.
- There is a `CompressedClustersFlat` variant. This cannot be used directly, but a `CompressedClusters` access structure can be created out of it simply via its constructor. Also a ROOT-seriazable version can be created like this, and then stored in a ROOT file.

The flat structure is quite simple:
- The memory backend is just a std::vector / array of chars (i.e. `char* memory` of size `size`)
- All the actual data (arrays of `CompressedClusters`'s ptrs) must live in this array of chars.
- At the beginning of the array, there must be space reserved for the `CompressedClustersFlat` struct, which is basically a header.
- Then, assuming a `CompressedClusters c` access structure, the header can be created via:
```
((CompressedClustersFlat*) memory)->set(size, &c);
```

I think, we could as well get rid of the `CompressedClustersROOT` and add a streamer for `CompressedClustersFlat` but I am not too familiar with that.

I tested this with the tpc-reco-workflow and also with @MichaelLettrich ' new encoder in the workflow and that works. However, the RootFileWriter crashes when it receives the Compressed Clusters Data (which is now sent in in the `CompressedClustersFlat` format). So I have to run the workflow like:
```
o2-tpc-reco-workflow --infile tpcdigits.root --output-type clusters,tracks,encoded-clusters,disable-writer
```
to disable the root-writing of compressed clusters.

That part should be fixed before we merge this. Perhaps we also need to work on the output of the Entropy Decoder for TPC.
In any case, I'd like to merge this soon, since some of the functionality is needed in more upcoming PRs (If needed, I could also split that off).

Could you have a look?
@matthiasrichter : Will you work on the RootTreeWriter part?
Or do we agree to just make the CompressedClustersFlat ROOT serializable? If so, I can also give it a try. I might come back with some questions then.
Where else do we have to adapt stuff (entropy decoding)?

shahor02 commented:
@davidrohr 

> However, the RootFileWriter crashes when it receives the Compressed Clusters Data (which is now sent in in the CompressedClustersFlat format).
I see a few problem there. The minor one is that the root IO requires for a class to have default c-tor. More serious one is that the root cannot store a single array with size >=1073741822 and this is exactly ballpark of expected TF size. If you could change it to a layout to something like
CompressedClustersFlat {

int size0 = 0;
...
int sizeN = 0;
char* ptr0 = nullptr; //[size0]
...
char* ptrN = nullptr; //[sizeN]
}
then it would be automatically persistent, w/o need for special streamer. This is very close to your CompressedClustersPtrs, you would just need to add to the pointers ``//[size..]`` comments and have these sizes (as ``int`` rather than ``size_t``!) as data members (could be also the members of a base class, i.e. something like ``CompressedClustersPtrs : public CompressedClustersCounters {..}`` would work.

davidrohr commented:
@shahor02 : This is how it was before actually. Matthias changed that to have a flat format in the ROOT streaming. Indeed, perhaps we do not need that anymore.

But still, it won't work that way. The TPC reco workflow is sending `CompressedClustersFlat`, which does not contain the pointers, but just some `size_t` entries with offsets. So it cannot be streamed with ROOT directly. One should
- convert to `CompressedClustersROOT`
- or add the ROOT annotation as you suggest + convert to `CompressedClusters`
- or add streamers for `CompressedClustersFlat`

shahor02 commented:
Yes, given that the CompressedClustersFlat is messageable, I would have only for the storage a version of CompressedClustersROOT with a ``pointer //[size]`` per data field and avoid storing large char array, this for sure will hit the root IO limitations. The price to pay will be that such an object, when read back from the file, is not guaranteed to have its pointers pointing on the single contiguous block of memory, so one would need to flatten before messaging (which does not mean large memory overhead as we would save on the large TBuffer used currently for reading single array from the file).

davidrohr commented:
OK, the point with ROOT io limitation is actually very important! In that case, we would remove the custom streamer and the `CompressedClustersRoot` completely, and revert the ROOT storage to the way as before, then already `CompressedClusters` is ROOT streamable (and messageable via ROOT streaming) while `CompressedClustersFlat` is trivially messagable but not ROOT-streamable. @matthiasrichter : Do you agree?

However, this still does not solve the problem that the RootTreeWriter cannot accept the message that contains the `CompressedClustersFlat` which is currently the output of the CATrackerSpec.
I guess the most trivial way to fix this, is to have two output types, one of the ROOT streamable type and one of the trivially messagable type. Then, one goes to the EntropyEncoderSpec (without any copy), and the other (a total clone of the output) goes to the RootTreeWriter, in case one wants to store the intermediate results. Does that make sense?

davidrohr commented:
The latest push contains different data types for the ROOT serializable compressed clusters, which are written and read to/from file, and for the trivially messageable flat version, that is used for direct shipment from the GPU tracker to the entropy encoder with zero-copy.
Both, running the full workflow `o2-tpc-reco-workflow -b --infile tpcdigits.root --output-type tracks,compressed-clusters,encoded-clusters` and creating encoded clusters from the stored compressed clusters from file `o2-tpc-reco-workflow -b --infile tpc-compclusters.root --output-type encoded-clusters --input-type compressed-clusters` works for me.

Could you have a look whether there are any reservations? Otherwise I would merge this and go ahead with filling the flat buffer for the tpc cluster native directly from the GPU as well.

shahor02 commented:
@davidrohr did not run myself, just going over your last modifications look ok.
